gcsio/src/main/java/com/google/cloud/hadoop/gcsio/DeleteFolderOperation.java

```diff
@@ -0,0 +1,237 @@
+package com.google.cloud.hadoop.gcsio;
+
+import static com.google.cloud.hadoop.gcsio.GoogleCloudStorageExceptions.createJsonResponseException;
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.api.client.googleapis.batch.json.JsonBatchCallback;
+import com.google.api.client.googleapis.json.GoogleJsonError;
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.client.http.HttpHeaders;
+import com.google.cloud.hadoop.util.ApiErrorExtractor;
+import com.google.common.base.Strings;
+import com.google.common.flogger.GoogleLogger;
+import com.google.storage.control.v2.DeleteFolderRequest;
+import com.google.storage.control.v2.StorageControlClient;
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentHashMap.KeySetView;
+import java.util.concurrent.LinkedBlockingQueue;
+import javax.annotation.Nonnull;
+
+public class DeleteFolderOperation {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  // Maximum number of times to retry deletes in the case of precondition failures.
+  private static final int MAXIMUM_PRECONDITION_FAILURES_IN_DELETE = 4;
+  private ApiErrorExtractor errorExtractor = ApiErrorExtractor.INSTANCE;
+  private GoogleCloudStorageOptions storageOptions;
+  private final KeySetView<IOException, Boolean> innerExceptions;
```

Comment:
Why is this a set? why is this called innerExceptions?

Reply:
This is just a list of exceptions that we get while deleting folders. InnerException because other usage of this set has been named innerException in GCSImpl class and thus maintains consistency.